File: node/runGovernor.js
const crypto = require("crypto")

function stableStringify(value) {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value)
  }

  if (Array.isArray(value)) {
    return `[${value.map(stableStringify).join(",")}]`
  }

  const keys = Object.keys(value).sort()
  const items = keys.map((key) => `${JSON.stringify(key)}:${stableStringify(value[key])}`)
  return `{${items.join(",")}}`
}

class RunGovernor {
  constructor({
    maxSteps = 50,
    maxToolCalls = 25,
    maxLlmCalls = 25,
    maxCostUsd = 1.0,
    maxWallClockSeconds = 300,
    maxRetriesPerStep = 2,
    loopSignatureWindow = 5,
    loopThreshold = 3,
    runId = null
  } = {}) {
    this.runId = runId || crypto.randomUUID()

    this.maxSteps = maxSteps
    this.maxToolCalls = maxToolCalls
    this.maxLlmCalls = maxLlmCalls
    this.maxCostUsd = maxCostUsd
    this.maxWallClockSeconds = maxWallClockSeconds
    this.maxRetriesPerStep = maxRetriesPerStep
    this.loopSignatureWindow = loopSignatureWindow
    this.loopThreshold = loopThreshold

    this.stepCount = 0
    this.toolCalls = 0
    this.llmCalls = 0
    this.costUsd = 0
    this.startedAt = Date.now()
    this.retriesByStep = {}
    this.recentToolSignatures = []
    this.trace = []
  }

  now() {
    return Date.now() / 1000
  }

  checkTimeout() {
    const elapsedSeconds = this.now() - (this.startedAt / 1000)
    if (elapsedSeconds > this.maxWallClockSeconds) {
      throw new Error("wall clock time limit exceeded")
    }
  }

  checkBudget() {
    if (this.costUsd > this.maxCostUsd) {
      throw new Error("cost limit exceeded")
    }
  }

  makeToolSignature(toolName, args = {}) {
    const raw = `${toolName}:${stableStringify(args)}`
    return crypto.createHash("sha256").update(raw).digest("hex")
  }

  checkLoopGuard(signature) {
    this.recentToolSignatures.push(signature)
    this.recentToolSignatures = this.recentToolSignatures.slice(-this.loopSignatureWindow)

    const repeats = this.recentToolSignatures.filter((s) => s === signature).length
    if (repeats >= this.loopThreshold) {
      throw new Error("loop guard triggered")
    }
  }

  appendTrace(event) {
    this.trace.push({
      run_id: this.runId,
      ...event
    })
  }

  recordStep(stepName = "default_step", summary = "") {
    this.checkTimeout()
    this.stepCount += 1

    if (this.stepCount > this.maxSteps) {
      throw new Error("step limit exceeded")
    }

    this.appendTrace({
      event_type: "step",
      step_index: this.stepCount,
      step_name: stepName,
      timestamp: this.now(),
      status: "ok",
      summary
    })
  }

  recordTool(toolName, args = {}, costUsd = 0, stepName = "default_step", summary = "") {
    this.checkTimeout()
    this.toolCalls += 1

    if (this.toolCalls > this.maxToolCalls) {
      throw new Error("tool call limit exceeded")
    }

    this.costUsd += costUsd
    this.checkBudget()

    const signature = this.makeToolSignature(toolName, args)
    this.checkLoopGuard(signature)

    this.appendTrace({
      event_type: "tool_call",
      step_index: this.stepCount,
      step_name: stepName,
      tool: toolName,
      arguments: args,
      tool_signature_hash: signature,
      cost_usd: costUsd,
      timestamp: this.now(),
      status: "ok",
      summary
    })

    return signature
  }

  recordLlmCall(model, tokens = 0, costUsd = 0, stepName = "default_step", summary = "") {
    this.checkTimeout()
    this.llmCalls += 1

    if (this.llmCalls > this.maxLlmCalls) {
      throw new Error("llm call limit exceeded")
    }

    this.costUsd += costUsd
    this.checkBudget()

    this.appendTrace({
      event_type: "llm_call",
      step_index: this.stepCount,
      step_name: stepName,
      model,
      tokens,
      cost_usd: costUsd,
      timestamp: this.now(),
      status: "ok",
      summary
    })
  }

  recordRetry(stepName, summary = "") {
    this.retriesByStep[stepName] = (this.retriesByStep[stepName] || 0) + 1

    if (this.retriesByStep[stepName] > this.maxRetriesPerStep) {
      throw new Error(`retry limit exceeded for step: ${stepName}`)
    }

    this.appendTrace({
      event_type: "retry",
      step_index: this.stepCount,
      step_name: stepName,
      retry_count: this.retriesByStep[stepName],
      timestamp: this.now(),
      status: "ok",
      summary
    })
  }

  summary() {
    return {
      run_id: this.runId,
      step_count: this.stepCount,
      tool_calls: this.toolCalls,
      llm_calls: this.llmCalls,
      cost_usd: Number(this.costUsd.toFixed(6)),
      elapsed_seconds: Number((this.now() - (this.startedAt / 1000)).toFixed(3))
    }
  }

  exportTrace() {
    return this.trace
  }
}

module.exports = { RunGovernor }

File: python/run_governor.py
import time
import json
import hashlib
import uuid
from typing import Any, Dict, List, Optional


class RunGovernor:
    def __init__(
        self,
        max_steps: int = 50,
        max_tool_calls: int = 25,
        max_llm_calls: int = 25,
        max_cost_usd: float = 1.0,
        max_wall_clock_seconds: int = 300,
        max_retries_per_step: int = 2,
        loop_signature_window: int = 5,
        loop_threshold: int = 3,
        run_id: Optional[str] = None,
    ) -> None:
        self.run_id = run_id or str(uuid.uuid4())

        self.max_steps = max_steps
        self.max_tool_calls = max_tool_calls
        self.max_llm_calls = max_llm_calls
        self.max_cost_usd = max_cost_usd
        self.max_wall_clock_seconds = max_wall_clock_seconds
        self.max_retries_per_step = max_retries_per_step
        self.loop_signature_window = loop_signature_window
        self.loop_threshold = loop_threshold

        self.step_count = 0
        self.tool_calls = 0
        self.llm_calls = 0
        self.cost_usd = 0.0
        self.started_at = time.time()
        self.retries_by_step: Dict[str, int] = {}
        self.recent_tool_signatures: List[str] = []
        self.trace: List[Dict[str, Any]] = []

    def _now(self) -> float:
        return time.time()

    def _check_timeout(self) -> None:
        elapsed = self._now() - self.started_at
        if elapsed > self.max_wall_clock_seconds:
            raise RuntimeError("wall clock time limit exceeded")

    def _check_budget(self) -> None:
        if self.cost_usd > self.max_cost_usd:
            raise RuntimeError("cost limit exceeded")

    def _stable_json(self, value: Any) -> str:
        return json.dumps(value or {}, sort_keys=True, separators=(",", ":"), default=str)

    def _make_tool_signature(self, tool_name: str, arguments: Optional[Dict[str, Any]]) -> str:
        raw = f"{tool_name}:{self._stable_json(arguments)}"
        return hashlib.sha256(raw.encode("utf-8")).hexdigest()

    def _check_loop_guard(self, signature: str) -> None:
        self.recent_tool_signatures.append(signature)
        self.recent_tool_signatures = self.recent_tool_signatures[-self.loop_signature_window:]

        repeats = self.recent_tool_signatures.count(signature)
        if repeats >= self.loop_threshold:
            raise RuntimeError("loop guard triggered")

    def _append_trace(self, event: Dict[str, Any]) -> None:
        event["run_id"] = self.run_id
        self.trace.append(event)

    def record_step(self, step_name: str = "default_step", summary: str = "") -> None:
        self._check_timeout()
        self.step_count += 1

        if self.step_count > self.max_steps:
            raise RuntimeError("step limit exceeded")

        self._append_trace(
            {
                "event_type": "step",
                "step_index": self.step_count,
                "step_name": step_name,
                "timestamp": self._now(),
                "status": "ok",
                "summary": summary,
            }
        )

    def record_tool(
        self,
        tool_name: str,
        arguments: Optional[Dict[str, Any]] = None,
        cost_usd: float = 0.0,
        step_name: str = "default_step",
        summary: str = "",
    ) -> str:
        self._check_timeout()
        self.tool_calls += 1

        if self.tool_calls > self.max_tool_calls:
            raise RuntimeError("tool call limit exceeded")

        self.cost_usd += cost_usd
        self._check_budget()

        signature = self._make_tool_signature(tool_name, arguments)
        self._check_loop_guard(signature)

        self._append_trace(
            {
                "event_type": "tool_call",
                "step_index": self.step_count,
                "step_name": step_name,
                "tool": tool_name,
                "arguments": arguments or {},
                "tool_signature_hash": signature,
                "cost_usd": cost_usd,
                "timestamp": self._now(),
                "status": "ok",
                "summary": summary,
            }
        )

        return signature

    def record_llm_call(
        self,
        model: str,
        tokens: int = 0,
        cost_usd: float = 0.0,
        step_name: str = "default_step",
        summary: str = "",
    ) -> None:
        self._check_timeout()
        self.llm_calls += 1

        if self.llm_calls > self.max_llm_calls:
            raise RuntimeError("llm call limit exceeded")

        self.cost_usd += cost_usd
        self._check_budget()

        self._append_trace(
            {
                "event_type": "llm_call",
                "step_index": self.step_count,
                "step_name": step_name,
                "model": model,
                "tokens": tokens,
                "cost_usd": cost_usd,
                "timestamp": self._now(),
                "status": "ok",
                "summary": summary,
            }
        )

    def record_retry(self, step_name: str, summary: str = "") -> None:
        self.retries_by_step[step_name] = self.retries_by_step.get(step_name, 0) + 1

        if self.retries_by_step[step_name] > self.max_retries_per_step:
            raise RuntimeError(f"retry limit exceeded for step: {step_name}")

        self._append_trace(
            {
                "event_type": "retry",
                "step_index": self.step_count,
                "step_name": step_name,
                "timestamp": self._now(),
                "status": "ok",
                "summary": summary,
                "retry_count": self.retries_by_step[step_name],
            }
        )

    def summary(self) -> Dict[str, Any]:
        return {
            "run_id": self.run_id,
            "step_count": self.step_count,
            "tool_calls": self.tool_calls,
            "llm_calls": self.llm_calls,
            "cost_usd": round(self.cost_usd, 6),
            "elapsed_seconds": round(self._now() - self.started_at, 3),
        }

    def export_trace(self) -> List[Dict[str, Any]]:
        return self.trace